prometheus/src/banks_with_commitments.rs

```diff
@@ -0,0 +1,67 @@
+use std::sync::{Arc, RwLock};
+
+use solana_runtime::{bank::Bank, bank_forks::BankForks, commitment::BlockCommitmentCache};
+
+use crate::utils::Metric;
+
+pub struct BanksWithCommitments {
+    pub finalized_bank: Arc<Bank>,
+    pub confirmed_bank: Arc<Bank>,
+    pub processed_bank: Arc<Bank>,
+}
+
+impl BanksWithCommitments {
+    pub fn new(
+        bank_forks: &Arc<RwLock<BankForks>>,
+        block_commitment_cache: &Arc<RwLock<BlockCommitmentCache>>,
+    ) -> Self {
+        let block_commitment_cache = block_commitment_cache.read().unwrap();
```

Comment:
### This can lead to a deadlock

This creates a lock guard that lives until the end of the scope, which means we hold the read lock on `block_commitment_cache` for the entire duration of the function, even though we no longer need it after we extract the slots.

Additionally, we take the `r_bank_forks` read lock later in this function, so then we hold two locks. Holding two locks simultaneously is dangerous if you don’t coordinate it very well among all other places that hold the same locks. The following scenario could happen:

* We take the read lock on `block_commitment_cache`.
* Some other thread takes a write lock on `bank_forks`.
* We want to take the read lock on `bank_forks`, but this blocks because the other thread holds a write lock.
* The other thread wants to take a write lock on `block_commitment_cache`, but that blocks because we hold the read lock.
* Deadlock!

I don’t want to audit the entire Solana codebase for all usages of those two locks to confirm that they are always acquired in the same order, and even if they are right now, there is no guarantee that nobody will mess it up in the future.

Fortunately, in this case we can easily avoid holding the locks at the same time:

```rust
let (finalized_slot, confirmed_slot, processed_slot) = {
    let block_commitment_cache = block_commitment_cache.read().unwrap();
    (
         block_commitment_cache.slot_with_commitment(solana_sdk::commitment_config::CommitmentLevel::Finalized),
         block_commitment_cache.slot_with_commitment(solana_sdk::commitment_config::CommitmentLevel::Confirmed),
         block_commitment_cache.slot_with_commitment(solana_sdk::commitment_config::CommitmentLevel::Processed),
    )
};
```
Now the lock guard will go out of scope at the `}`, before we acquire the lock on `bank_forks`.

### Fixing the deadlock may lead to inconsistencies

So now we can avoid the deadlock, but there is another problem. The bank forks and commitment cache really belong together, the commitment cache refers to slots that the bank holds. Assuming the same order as above (we read the commitment cache, then the bank forks; some other thread writes the bank forks, then updates the cache), I can imagine the following would happen:

* We read the commitment cache and find slot _x_ is the finalized slot.
* The other thread just received a new confirmed block, locks the bank forks, and puts the new block in. The new finalized slot is slot _x + 1_, and the bank for slot _x_ gets evicted.
* The other thread then updates the commitment cache to reflect the new slots, so that the commitment cache would return slot _x + 1_ for level finalized.
* We try to get the bank for slot _x_, but by now that slot has been evicted from the bank forks.

Now I haven’t looked into detail in how the bank forks work, and when things get evicted, and how the commitment config and bank forks are updated, so this is speculation on my part. But I wouldn’t be surprised at all if this scenario is the cause of https://github.com/solana-labs/solana/issues/11078 linked below. (Which was closed because it was “fixed” by this workaround that we also apply here.)

With this insight, there is an alternative workaround: instead of falling back to to the root slot (which is the same as the one with finalized commitment level), if we try to get a slot but it’s not there, that means the bank forks have changed with respect to the commitment cache since we read the commitment cache, and we could start over. We read both right after one another, so this case should be rare, and if there is say even a 10% probability that we need a restart, then the probability that we still haven’t succeeded after _n_ attempts goes to 0 rapidly as _n_ grows ... assuming that these events are really independent — in practice there is no guarantee that this process ever terminates.

So as a hack for rare situations, I think the current workaround is good enough.

### Proper solution (Rant)

I think it’s a design mistake on Solana’s part to put these two things in different locks. It creates all kinds of very subtle problems, and as https://github.com/solana-labs/solana/issues/11078 demonstrates, they are not just hypothetical. To me it looks like `BankForks` and `CommitmentCache` should really be merged into one thing, or at the very least they should be in a `RwLock<(BankForks, CommitmentCache)>`, instead of having their own lock.

(Even nicer of course would be the “private producer + atomic swap” approach that we use in Solido, instead of an `RwLock`, such that reads can proceed even if a write is in progress, and such that writes don’t need to block on concurrent reads.)

But let’s not go down this rabbit hole, the Solana codebase is the way it is, and we have to deal with it.

Reply:
Thanks for the detailed explanation. Indeed there's contention on these two locks, should be more careful when doing `read()` more than once. I changed the guard to limit its scope as you said. It would be great if Solana changed this the way we do it in Solido 😬.